COMPONENT-GUIDELINES.md

```diff
@@ -0,0 +1,293 @@
+
+**Actions**
+
+*Coming soon*
+
+# Guidelines & Patterns
+
+## General
+
+### Component Scope
+
+Create components to address specific use cases whenever possible. For example, when a user subscribes to a Github webhook to listen for “star” activity events can be generated when users star or unstar a repository. The “New Star” source filters events for only new star activity so the user doesn’t have to.
+
+There may be cases where it's valuable to create a generic component that provides users with broad lattitude (e.g., see the [custom webhook](components/github/sources/custom-webhook-events) event source for Github). However, as a general heuristic, we found that tightly scoped components are easier for users to understand and use.
+
+### Required Metadata
+
+Registry components require a unique key and version, and a friendly name and description. Action components require a `type` field to be set to `action` (sources will require a type to be set in the future).
+
+### Folder Structure
+
+Registry components are organized by app in the `components` directory of the `pipedreamhq/pipedream` repo. 
+
+```
+/components
+	/[app_name_slug]
+		/[app_name_slug].app.js
+		/actions
+			/[action_name_slug]
+				/[action_name_slug].js
+		/sources
+			/[source_name_slug]
+				/[source_name_slug].js
+```
+
+- The name of each app folder corresponds with the name slug for each app
+- The app file should be in the root of the app folder (e.g., `../components/[app_slug]/[app_slug].app.js`)
+- Components for each app are organized into `/sources` and `/actions` subfolders
+- Each component should be placed in its own subfolder (with the name of the folder and the name of the `js` file equivalent to the slugified component name). For example, the path for the "Search Mentions" source for Twitter is `../components/twitter/sources/search-mentions/search-mentions.js`.
+
+You can explore examples in the [components directory](components).
+
+## Promoting Reusability
+
+### App Files
+
+App files contain components that declare the app and include prop definitions and methods that may be reused across components. App files should adhere to the following naming convention:  `[app_name_slug].app.js`. If an app file does not exist for your app, please [reach out](https://pipedream.com/community/c/dev/11).
+
+#### Prop Definitions 
+
+Whenever possible, reuse existing [prop definitions](https://github.com/PipedreamHQ/pipedream/blob/master/COMPONENT-API.md#prop-definitions-example). If specific elements (e.g., default value, optional/required state) for a prop definition does not fit with the current use case, those values can be overridden by redefining the values for specific keys in the source. If a prop definition does not exist and you are adding an app-specific prop that may be reused in future components, add it as a prop definition to the app file. Prop definitions will also be surfaced for apps the Pipedream marketplace. 
```

Comment:
"those values can be overridden by redefining the values for specific keys in the source" -> might be worth an example?